docs/flip/1009-optimizer-api.md

```diff
@@ -0,0 +1,471 @@
+    (biases_traversal, flax.optim.GradientDescent(learning_rate=0.1)),
+    (not_biases_traversal, flax.optim.GradientDescent(learning_rate=0.05)),
+)
+```
+
+Note how we first define a traversal that selects parameters based on their
+path (which is the concatenation of module scopes and variable name), and then
+create a `MultiOptimizer` that binds a different optimizer for each of these
+separate traversals.
+
+Optax has recently implemented `optax.masked()` that can be used for specifying
+gradient transformations that only applied to a subset of the gradients:
+
+```python
+def flattened_traversal(fn):
+  def mask(data):
+    flat = traverse_util.flatten_dict(data)
+    return traverse_util.unflatten_dict({k: fn(k, v) for k, v in flat.items()})
+  return mask
+
+tx = optax.chain(
+    optax.masked(optax.sgd(learning_rate=0.1),
+                 mask=flattened_traversal(lambda path, _: path[-1] == 'bias')),
+    optax.masked(optax.sgd(learning_rate=0.05),
+                 mask=flattened_traversal(lambda path, _: path[-1] != 'bias')),
+)
+```
+
+## Train State
+[Train State]: #train-state
+
+In Flax it is common to hand around a `TrainState` object that can then be
+used for checkpointing. This simplifies the above [Optax training step] a bit by
+reducing the number of arguments and getting rid of the `static_argnums`.
+
+We can define an `Optimizer` dataclass that wraps the common pattern of updating
+the optimizer state and parameters, and keeps `params` that are to be optimized
+separate from `model_state` that is not touched by the optimizer:
+
+```python
+class Optimizer(flax.struct.PyTreeNode):
+  params: flax.core.FrozenDict[str, Any]
+  model_state: flax.core.FrozenDict[str, Any]
+  tx: optax.GradientTransformation = flax.struct.field(pytree_node=False)
+  opt_state: optax.OptState
+
+  def update(self, grads, new_model_state):
+    updates, new_opt_state = self.tx.update(
+        grads, self.opt_state, self.params)
+    new_params = optax.apply_updates(self.params, updates)
+    return self.replace(
+        params=new_params,
+        opt_state=new_opt_state,
+        model_state=new_model_state,
+    )
+
+  @classmethod
+  def create(cls, variables, tx):
+    model_state, params = variables.pop('params')
+    opt_state = tx.init(params)
+    return cls(
+        params=params,
+        model_state = model_state,
+        tx=tx,
+        opt_state=opt_state,
+    )
+```
+
+Users can then compose their own data class containing this optimizer, the
+model's `.apply()` function and arbitrary additional data:
+
+```python
+class TrainState(flax.struct.PyTreeNode):
+  step: int
+  apply_fn: Callable = flax.struct.field(pytree_node=False)
+  optimizer: Optimizer
+```
+
+With this the [Optax Training Step] becomes:
+
+```python
+@jax.jit
+def train_step(state, inputs, labels):
+
+  def loss_fn(params):
+    outputs, new_model_state = state.apply_fn(
+        state.optimizer.model_state.copy({'params': params}),
```

Comment:
`{'params': params, **state.optimizer.model_state}` does the same thing, I think? And doesn't require using special Flax APIs

Reply:
Updated.